# CSOmap
======installation=====

We provide both matlab version and R version, please refer to the corresponding README.md

We highly recommend using the matlab version if possible. It is the version that we used in our paper "Reconstruction of cell spatial organization from single-cell RNA sequencing data based on ligand-receptor mediated self-assembly".

The R version includes the main algorithm, but it does not provide figures or statistical tests.

We compressed the demo dataset via gzip, please decompress the .gz files before running any scripts.

We provided a demo dataset as an example of data preparation. You can refer to the README file in both matlab and R version to find out how to run it. You can also run your own datasets.

We provided both human and mouse ligand receptor datasets in the demo/ folder. 

Human LR dataset reference: 

1) Ramilowski, J., Goldberg, T., Harshbarger, J. et al. A draft network of ligand–receptor-mediated multicellular signalling in human. Nat Commun 6, 7866 (2015). https://doi.org/10.1038/ncomms8866

2) Chen, L., Flies, D. Molecular mechanisms of T cell co-stimulation and co-inhibition. Nat Rev Immunol 13, 227–242 (2013). https://doi.org/10.1038/nri3405

Mouse LR dataset reference: 

1) Halpern, K., Shenhav, R., Massalha, H. et al. Paired-cell sequencing enables spatial gene expression mapping of liver endothelial cells. Nat Biotechnol 36, 962–970 (2018). https://doi.org/10.1038/nbt.4231

